src/content/post/2024-04-05-rdf-shacl-and-kuzu.md

```diff
@@ -0,0 +1,343 @@
+9    http://kuzu.io/rdf-ex#Karissa  http://www.w3.org/1999/02/22-rdf-syntax-ns#type   http://kuzu.io/rdf-ex#student
+10      http://kuzu.io/rdf-ex#Adam  http://www.w3.org/1999/02/22-rdf-syntax-ns#type   http://kuzu.io/rdf-ex#student
+11   http://kuzu.io/rdf-ex#faculty  http://www.w3.org/2000/01/rdf-schema#subClassOf    http://kuzu.io/rdf-ex#person
+12   http://kuzu.io/rdf-ex#student  http://www.w3.org/2000/01/rdf-schema#subClassOf    http://kuzu.io/rdf-ex#person
+13     http://kuzu.io/rdf-ex#Zhang  http://www.w3.org/1999/02/22-rdf-syntax-ns#type   http://kuzu.io/rdf-ex#faculty
+```
+
+## Specify SHACL shape constraints
+
+SHACL allows us to define constraints on the RDF graph. To demonstrate how this works, consider this rather strict
+constraint:
+
+```turtle
+@prefix sh:   <http://www.w3.org/ns/shacl#> .
+@prefix xsd:  <http://www.w3.org/2001/XMLSchema#> .
+@prefix kz: <http://kuzu.io/rdf-ex#> .
+
+kz:PersonShape
+    a sh:NodeShape ;
+    sh:targetClass kz:student ;
+    sh:property [
+        sh:path kz:name ;
+        sh:minLength 5
+    ] .
+```
+
+The above lines check that the `name` property of a `student` resource has a minimum length of 5 characters.
+
+## Validate RDF data against SHACL shapes
+
+We can use the `pySHACL` library to validate the RDF data against SHACL shapes. The shape constraint
+from above is read into Python code and used to validate the RDF graph.
+
+```py
+import pyshacl
+# `shapes_graph` is the SHACL constraint from above specified in Turtle format
+
+results = pyshacl.validate(
+    graph,
+    shacl_graph = shapes_graph,
+    data_graph_format = "json-ld",
+    shacl_graph_format = "ttl",
+    inplace = False,
+    serialize_report_graph = "ttl",
+    debug = False,
+)
+```
+
+Sure enough, the validation fails in this case because in our data, the name of the student `Adam`
+has only 4 characters. The validation report is output so that we can see the constraint violation.
+
+```turtle
+Validation Report
+Conforms: False
+Results (1):
+Constraint Violation in MinLengthConstraintComponent (http://www.w3.org/ns/shacl#MinLengthConstraintComponent):
+        Severity: sh:Violation
+        Source Shape: [ sh:minLength Literal("5", datatype=xsd:integer) ; sh:path kz:name ]
+        Focus Node: <http://kuzu.io/rdf-ex#Adam>
+        Value Node: Literal("Adam")
+        Result Path: kz:name
+        Message: String length not >= Literal("5", datatype=xsd:integer)
+```
+
+The violation can easily be fixed by setting the minimum length of the `name` property to a smaller value, such as 2.
+When this is done, the validation passes.
+
+```turtle
+Validation Report
+Conforms: True
+```
+
+## Visualize the RDF graph in Kùzu Explorer
+
+When building and validating RDF graphs, the ability to get visual feedback is quite useful. Kùzu
+Explorer is a web-based interface that allows you to visualize RDF graphs and query them using Cypher (no knowledge of SPARQL required).
```

Comment:
I think we use KuzuExplorer instead of "Kuzu Explorer", so let's fix that throughout the blog post.

Reply:
I think the term KùzuExplorer is a little ungainly and repetitive, and I've updated [the docs](https://docs.kuzudb.com/visualization/) to state them as separate words. From a branding perspective, it's similar to a lot of other tools (e.g., Neo4j Desktop), where the company/database name is kept separate from the tool. As long as we're consistent, I think the separated words reads better and we can also use the term "Explorer" where needed.